flocker/node/agents/blockdevice.py

```diff
@@ -340,3 +504,113 @@ def get_device_path(self, blockdevice_id):
+    :param IBlockDeviceAPI block_device_api: The block device API that will be
+        called upon to perform block device operations.
+    :ivar FilePath mountroot: The directory where block devices will be
+        mounted.
+    """
+    hostname = field(type=unicode, mandatory=True)
+    block_device_api = field(mandatory=True)
+    mountroot = field(type=FilePath, initial=FilePath(b"/flocker"))
+
+    def discover_local_state(self):
+        """
+        Find all block devices that are currently associated with this host and
+        return a ``NodeState`` containing only ``Manifestation`` instances and
+        their mount paths.
+        """
+        volumes = self.block_device_api.list_volumes()
+
+        manifestations = [_manifestation_from_volume(v)
+                          for v in volumes
+                          if v.host == self.hostname]
+
+        paths = {}
+        for manifestation in manifestations:
+            dataset_id = manifestation.dataset.dataset_id
+            mountpath = self._mountpath_for_manifestation(manifestation)
+            paths[dataset_id] = mountpath
+
+        state = NodeState(
+            hostname=self.hostname,
+            running=(),
+            not_running=(),
+            manifestations=manifestations,
+            paths=paths
+        )
+        return succeed(state)
+
+    def _mountpath_for_manifestation(self, manifestation):
+        """
+        Calculate a ``Manifestation`` mount point.
+
+        :param Manifestation manifestation: The manifestation of a dataset that
+            will be mounted.
+        :returns: A ``FilePath`` of the mount point.
+        """
+        return self.mountroot.child(
+            manifestation.dataset.dataset_id.encode("ascii")
+        )
+
+    def calculate_necessary_state_changes(self, local_state,
+                                          desired_configuration,
+                                          current_cluster_state):
+        potential_configs = list(
+            node for node in desired_configuration.nodes
+            if node.hostname == self.hostname
+        )
+        if len(potential_configs) == 0:
+            this_node_config = Node(hostname=None)
+        else:
+            [this_node_config] = potential_configs
+
+        configured_dataset_ids = set(
+            manifestation.dataset.dataset_id for manifestation in
+            this_node_config.manifestations.values()
+            # Don't create deleted datasets
+            if not manifestation.dataset.deleted
+        )
+
+        local_dataset_ids = set(
+            manifestation.dataset.dataset_id for manifestation in
+            local_state.manifestations
+        )
+
+        manifestations_to_create = set(
+            this_node_config.manifestations[dataset_id] for dataset_id in
+            configured_dataset_ids.difference(local_dataset_ids)
+        )
+
+        # TODO check for non-None size on dataset; cannot create block devices
```

Comment:
I think this should be enforced on configuration side (what are you going to do here?), see FLOC-1579.